storage/src/mmr/mutable.rs

```diff
@@ -0,0 +1,385 @@
+    pub fn operation_count(&self) -> u64 {
+        self.log.len() as u64
+    }
+
+    /// Updates `key` to have value `value`.  If the key already has this same value, then this is a
+    /// no-op.
+    pub fn update(&mut self, hasher: &mut H, key: K, value: V) {
+        let new_location = self.log.len();
+
+        // Update the snapshot.
+        if let Some(location) = self.snapshot.get_mut(&key) {
+            let last_value = match self.log[*location].op_type {
+                Type::Deleted => panic!("deleted key should not be in snapshot: {}", key),
+                Type::Update(ref v) => v,
+            };
+            if value == *last_value {
+                // Trying to assign the same value is a no-op.
+                return;
+            }
+            *location = new_location;
+        } else {
+            self.snapshot.insert(key.clone(), new_location);
+        }
+
+        let op = Type::Update(value.clone());
+        self.update_work(hasher, key, op);
+    }
+
+    /// Update the log and operations MMR with the given key and operation.
+    fn update_work(&mut self, hasher: &mut H, key: K, op_type: Type<V>) {
+        let op = Operation { key, op_type };
+
+        // Update the ops MMR.
+        let digest = Self::op_digest(hasher, &op);
+        self.ops.add(hasher, &digest);
+
+        // Append the operation to the log.
+        self.log.push(op);
+    }
+
+    /// Delete `key` and its value from the store. Deleting a key that already has no value is a
+    /// no-op.
+    pub fn delete(&mut self, hasher: &mut H, key: K) {
+        // Remove the key from the snapshot.
+        if self.snapshot.remove(&key).is_none() {
+            return;
+        }
+
+        self.update_work(hasher, key, Type::Deleted);
+    }
+
+    /// Return the root hash of the mutable MMR.
+    pub fn root(&self, hasher: &mut H) -> H::Digest {
+        self.ops.root(hasher)
+    }
+
+    /// Return the size of the underlying MMR.
+    ///
+    /// This value will be the position of the next leaf to be added to the underlying MMR, and can
+    /// be used to retrieve a starting position value for generating a proof over the next batch of
+    /// operations.
+    pub fn size(&self) -> u64 {
+        self.ops.size()
+    }
+
+    /// Generate and return:
+    ///  1. a proof of all operations applied to the store in the range starting at (and including)
+    ///     location `start_location`, and ending at the first of either:
+    ///         - the last operation performed, or
+    ///         - the operation `max_ops` from the start.
+    ///  2. the operations corresponding to the leaves in this range.
+    pub async fn proof_to_tip(
+        &self,
+        start_loc: u64,
+        max_ops: u64,
+    ) -> Result<(Proof<K, V, H>, Vec<Operation<K, V>>), Error> {
+        let start_pos = leaf_num_to_pos(start_loc);
+        let end_pos_last = self.ops.last_leaf_pos().unwrap();
```

Comment:
If called on an empty MMR, this would now panic right (as we don't start with anything)?

Reply:
Yes implementation assumes the user has provided a valid start_loc (meaning the mmr would never be empty).